driver/gather_driver.hpp

```diff
@@ -0,0 +1,389 @@
+                         "int");
+    inflags.AddInputFlag(
+        "contiguous", 'C', "1", "Tensor is contiguous (1) or not (0) (Default=1)", "int");
+    inflags.AddTensorFlag(
+        "input_shape", 'P', "2x3x5x5", "The shape of the input tensor (Default=2x3x5x5)");
+    inflags.AddTensorFlag(
+        "indices_shape", 'I', "2x3x3x3", "The shape of the indices tensor (Default=2x3x3x3)");
+    inflags.AddInputFlag(
+        "mode", 'm', "gather", "Gather Mode (gather, gatherv2, gathernd) (Default=gather)", "str");
+    inflags.AddInputFlag(
+        "dim", 'D', "0", "The dimension in params to gather indices from (Default=0)", "int");
+    inflags.AddInputFlag(
+        "batch_dims", 'B', "0", "The number of batch dimensions (Default=0)", "int");
+    inflags.AddInputFlag("iter", 'i', "10", "Number of Iterations (Default=10)", "int");
+    inflags.AddInputFlag("verify", 'V', "1", "Verify Each Layer (Default=1)", "int");
+    inflags.AddInputFlag("time", 't', "0", "Time Each Layer (Default=0)", "int");
+    inflags.AddInputFlag(
+        "wall", 'w', "0", "Wall-clock Time Each Layer, Requires time == 1 (Default=0)", "int");
+
+    return miopenStatusSuccess;
+}
+
+template <typename Tgpu, typename Tref, typename Tindex>
+int GatherDriver<Tgpu, Tref, Tindex>::AllocateBuffersAndCopy()
+{
+    uint32_t ctx = 0;
+
+    size_t input_sz   = GetTensorSpace(inputTensor);
+    size_t output_sz  = GetTensorSpace(outputTensor);
+    size_t indices_sz = GetTensorSpace(indicesTensor);
+
+    // GPU allocation
+    input_dev   = std::unique_ptr<GPUMem>(new GPUMem(ctx, input_sz, sizeof(Tgpu)));
+    output_dev  = std::unique_ptr<GPUMem>(new GPUMem(ctx, output_sz, sizeof(Tgpu)));
+    indices_dev = std::unique_ptr<GPUMem>(new GPUMem(ctx, indices_sz, sizeof(Tindex)));
+
+    // GPU host allocation
+    input   = std::vector<Tgpu>(input_sz, static_cast<Tgpu>(0));
+    output  = std::vector<Tgpu>(output_sz, static_cast<Tgpu>(0));
+    indices = std::vector<Tindex>(indices_sz, static_cast<Tindex>(0));
+
+    // CPU allocation
+    outputHost = std::vector<Tref>(output_sz, static_cast<Tref>(0));
+
+    for(size_t i = 0; i < input_sz; i++)
+    {
+        input[i] = prng::gen_A_to_B(static_cast<Tgpu>(0.0), static_cast<Tgpu>(1.0));
+    }
+
+    for(size_t i = 0; i < indices_sz; i++)
+    {
+        auto input_shape = miopen::deref(inputTensor).GetLengths();
+        indices[i] =
+            prng::gen_A_to_B(static_cast<Tindex>(0), static_cast<Tindex>(input_shape[dim]));
+    }
+
+    if(indices_dev->ToGPU(GetStream(), indices.data()) != 0)
+    {
+        std::cerr << "Error copying (indices) to GPU, size: " << indices_dev->GetSize()
+                  << std::endl;
+        return miopenStatusInternalError;
+    }
+
+    if(input_dev->ToGPU(GetStream(), input.data()) != 0)
+    {
+        std::cerr << "Error copying (input) to GPU, size: " << input_dev->GetSize() << std::endl;
+        return miopenStatusInternalError;
+    }
```

Comment:
I noticed that `output_dev` isn't being transferred to the GPU. Could this potentially cause any issues? I understand that `output_dev`'s values are just `NaN` or empty values, which are filled later after running the op, but it seems that, in other operations, output tensors are still transferred to the GPU beforehand.

Reply:
It checks for errors in transferring data between the CPU and GPU, which may help detect errors beforehand, but I don't think it would cause any issues.